compiler/rustc_passes/src/check_attr.rs

```diff
@@ -2052,6 +2083,90 @@ impl CheckAttrVisitor<'_> {
             errors::Unused { attr_span: attr.span, note },
         );
     }
+
+    fn check_proc_macro(&self, hir_id: HirId, target: Target, kind: ProcMacroKind) {
```

Comment:
This check breaks potentially existing code.

We don't currently require that the function signature is *equal* to `fn(TokenStream) -> TokenStream`, but instead *monomorphizes* to `fn(TokenStream) -> TokenStream` after some substitution of generic arguments.

That is, this code is valid as far as I can tell:
```
#[proc_macro]
pub fn foo<T>(t: T) -> TokenStream {
  // ...
}
```

... but would fail with your check here.

---

You probably need to be a bit more general about argument compatibility here. I would:
1. create an inference context, 
2. create fresh substitutions for the function you're checking,
3. instantiate the signature with those substitutions,
4. turn the signature into a fn-pointer type and equate it against an expected signature of `fn(TokenStream) -> TokenStream` to do inference. 

You probably want to do this with an `ObligationCtxt` as well, and also report any errors if the signature equality created any where-clause obligations which are impossible to satisfy.

Reply:
I've gone with the `DeepRejectCtxt` approach instead.